app/assets/javascripts/settings/ldap-conn-test.js

```diff
@@ -18,6 +18,56 @@ $(function () {
     $('#ldap_conn_message').html(msg)
   }
 
+  function ldapConnectionTestPart1 () {
+    var file = $('#dex_connector_ldap_certificate').val()
+    if(!file) {
+      ldapConnectionTestPart2( $('#dex_connector_ldap_current_cert').val() )
+      return;
+    }
+    var fileInput = $('dex_connector_ldap_certificate')[0];
+    var reader = new FileReader();
+
+    reader.onload = function(e) {
+      ldapConnectionTestPart2(reader.result);
+    }
+
+    reader.readAsText(fileInput.files[0]);	
+  }
+
+  function stripCert (cert) {
+    cert = cert.replace("-----BEGIN CERTIFICATE-----\n",'');
+    cert = cert.replace("\n-----END CERTIFICATE-----",'');
+    return cert;
+  }
+
+  function ldapConnectionTestPart2 (cert) {
+    var host     = $('#dex_connector_ldap_host').val()
+    var port     = $('#dex_connector_ldap_port').val()
+    var startTLS = $('#dex_connector_ldap_start_tls_true').parent().hasClass('active')
+    // Flip logic of #anonymous_bind_toggle since expanded div means anonymous binding is 'false'
+    var anonBind = ($('#anonymous_bind_toggle').attr('aria-expanded') === 'false')
+    var dn       = $('#dex_connector_ldap_bind_dn').val()
+    var pass     = $("#dex_connector_ldap_bind_pw[required='required']").val();
+    var baseDN   = $('#dex_connector_ldap_user_base_dn').val()
+    var filter   = $('#dex_connector_ldap_user_filter').val()
+    var paramArray = anonBind ? [host, port, cert, baseDN, filter] : [host, port, cert, dn, pass, baseDN, filter]
+
+    if (noneEmpty(paramArray)) { // Check if form inputs are filled in
+      $.post('/settings/ldap_test', {
+        host: host, port: port, start_tls: startTLS, cert: stripCert(cert),
```

Comment:
One improvement that could be done in this regard was to instead of declaring a variable for each value, we could declare a single variable as an object and add the values to it. Something like

```js
var params = {
  host: $('#dex_connector_ldap_host').val(),
  port: $('#dex_connector_ldap_port').val(), 
  //...
}
```
This would require a change to the `paramArray` assignment like

```js
if (anonBind) {	
  paramArray = [params.host, params.port, params.cert, params.baseDN, params.filter]	
} else { //... } 
```

I'm suggesting to revert to the if/else block because the line would be too long with the addition of the `params` prefix.

And finally just 

```js
$.post('/settings/ldap_test', params)
```

But this is just cosmetics. Since we don't have eslint to enforce any style, almost everything is fine.

Reply:
The `params` could also be extracted to its own function too. That would reduce the amount of lines of this method. Another possible enhancement. 

And don't worry too much about the codeclimate complains. We don't have an eslint to enforce stuff, so from time to time it's fine to ignore those kind of complains.

Reply:
The function as written is designed to give an error if some field is not filled in. If that is viewed as the goal it does it in an efficient way with a relatively low number of lines of code. It is quite clear how it works. I would not have written it this way myself; the original code here was written by Daniel Orf, but I do think it is accomplishing the purpose in an acceptable fashion.

I don't think the UX flow is good. What really should happen is that fields are marked as required ( typically with a red star ), and that when you attempt to save without any of the fields present, that the fields that are empty are set with a red border on the entire field and that the scroll automatically scrolls up to the first of the fields that has not been entered. That is the common pattern that should be followed for a good user experience.

That needs to happen in the near future. When that happens this code you are suggesting to optimize would go away because it will be deleted and replaced with something else entirely. For this reason optimizing what we have here is premature and should not be done and is a waste of developer time.

Reply:
I agree that denoting required fields with a red star/border should be added in a separate bug.